Replace metadata + fetch + cache code

zkat opened the pull request:
## Summary

This PR replaces everything in the current codebase having to do with "give me the metadata for this package" and "extract this package to this place" with a single library, [`pacote`](https://npm.im/pacote) which abstracts away the whole thing. This includes all the caching code, since `pacote` handles all of that internally through [`cacache`](https://npm.im/cacache)

This means the following bits are being entirely replaced:

* `fetchPackageMetadata`
* `lib/cache`
* `mapToRegistry`
* All calls to `cache.*` in various parts of the codebase
* All bits in `lib/doctor` having to do with cache verification (`cacache` has its own verifier this can call)
* others?

On the user side, the only `breaking` change will be that the cache structure will have completely changed, and some incidentals of cache management may end up getting tweaked, but by the time this is done, it should be mostly compatible.

## Status

Initial registry dep support in `pacote` is sketched out, and integration can begin. Expect lots of loose ends and a lot of back-and-forth between the libraries as the API is sussed out in more detail.

### UPDATE (2017-04-17):

This branch is now usable. It needs some fixes, and the test suite is still failing due to compatibility issues and probably-minor issues. It is not production ready, but it runs and it works. All the work on it is now going into getting the npm test suite passing, at which point we'll be able to do a final review + merge.

### UPDATE (2017-03-14):

Currently, [cacache](https://github.com/zkat/cacache) is production-ready, fast, and really damn reliable.

Most work is currently focused on [pacote](https://github.com/zkat/pacote), and finishing up and testing that all of its package types actually work. This should be done soon, and then this branch can start moving again. The only major component left is [git support](https://github.com/zkat/pacote/pull/69). `pacote@1` will be tagged once that's done, and further versions added during integration and test fixing for the npm CLI test suite (which I haven't run and am kinda terrified about)

### I want to help!

Cool! Head on over to the [pacote](https://github.com/zkat/pacote) and [cacache](https://github.com/zkat/cacache) git repositories and look in the issues for `help-wanted` and `starter` bits and comment there if you want to help with those.

iarna commented:
@zkat If we can get this patch as is, passing tests, I think I'd like to merge it before we move on to further integration. What do you think?

zkat commented:
@iarna https://github.com/zkat/pacote/milestone/2 There is now a dedicated milestone for `pacote@0.1.0` defined as "whatever we need in order to replace the current registry-install stack with pacote".

iarna committed:
install: Replace post install module list with summary report

thomaschaaf commented on doc/misc/npm-config.md:
```diff
@@ -628,6 +628,14 @@ The registry you want to send cli metrics to if `send-metrics` is true.
+### offline
+
+* Default: false
+* Type: Boolea
```
Boolean

zkat replied:
lol nice catch. Thank you.

iarna committed:
install: Print warnings before summary

iarna committed:
tests: Update tests to support new output formats

iarna committed:
dedupe: find-dupes should output _something_

npm find-dupes is now an alias for `npm dedupe --dry-run --parseable`

iarna committed:
build: Only output symlink info when doing human output

iarna committed:
ls: Remove blank lines from `npm ls --parseable`

zkat committed:
test: fixing standard fails

iarna committed:
install/deps: Remove needlessly async calls to npa

Credit: @iarna

zkat committed:
fix(save): calculate the version to save for new registry deps (#16414)

iarna committed:
install: Fetch as much registry metadata we can at once (#16422)

zkat committed:
pacote@2.7.0

zkat committed:
fix(tests): getting test suite working again

zkat committed:
feat(intall): use pacote for package downloads

zkat committed:
fix(shrinkwrap): make sure resolved is there

zkat committed:
bring stuff closer to old version

zkat committed:
refactor(shrinkwrap): refactor and promisify inflate-shrinkwrap.js

zkat committed:
fix(shrinkwrap): include and use bundle information in shrinkwrap

zkat committed:
fix(noise): stop the gentlyRm spam omg

zkat committed:
fix(tests): get.js is no longer used for package downloads and is not cached

zkat committed:
fix(tests): update retry-on-stale-cache to use --prefer-offline

zkat committed:
fix(test): updated splat-with-only-prerelease-latest test

zkat committed:
pacote bump

zkat committed:
feat(opts): save installs by default

BREAKING CHANGE: you need to use --no-save to skip saves now;

zkat committed:
feat(logger): added `notice` loglevel

zkat committed:
feat(lockfile): add package-lock.json support

This implements https://github.com/npm/npm/pull/16441 as well as the auto-save behavior it needs in order to be kept up to date.

zkat committed:
FLOATING PATCH: directly patch fstream-npm

remove this patch once https://github.com/npm/fstream-npm/pull/27 lands

zkat committed:
feat(clean): hide npm cache clean behind --force

zkat committed:
fix(cache): reject any attempt to remove individual packages

zkat committed:
meta: dogfood package-lock.json

zkat committed:
fix(extract): cut down on EBUNDLEOVERRIDE spam

zkat committed:
fix(shrinkwrap): should still work if lockfile has no deps

zkat committed:
refactor(extract): covert extract.js to Promises

zkat committed:
feat(shrinkwrap): always shrinkwrap devDeps, even if only devDeps are saved

iarna committed:
fix(shrinkwrap): force installs to use the shrinkwrap tree

zkat committed:
trying to fix ci

zkat committed:
missed more spots about shrinkwrap

zkat committed:
fix standard bits

zkat committed:
test suite fixes

zkat committed:
disable shrinkwrap for self-install test

zkat committed:
npm it test was clobbering package-lock.json

zkat committed:
fix(shrinkwrap): write lockfile to the tree root

zkat committed:
fix(save): pay attention to save type

zkat committed:
generate the damn thing all the time

zkat committed:
fix(cache): corrupted cache entries properly detected and fixed, again

zkat committed:
fix(pick-manifest): requested-versions with spaces are now handled properly

zkat committed:
fix(pick-manifest): semver specs using = work again

zkat committed:
fix(auth): authentication fixed on custom registries with specific paths

zkat committed:
fix(extract): stop trusting package.json in the tarball

zkat committed:
feat(git): git repos with prepare scripts are now built on install

zkat committed:
fix(version): properly detect whether lockfile or shrinkwrap

zkat committed:
fix(git): only use -c longpaths on win32

zkat committed:
feat(log): moved new loglevels to npmlog proper

zkat committed:
feat(indent): package.json and package locks preserve indentaton

zkat committed:
add detect-indent to bundle

zkat committed:
fix(shrinkwrap): npm shrinkwrap should create a shrinkwrap by default, not a pkglock

zkat committed:
fix(tests): shrinkwrap always includes the full tree

zkat committed:
fix(test): fixfix

zkat committed:
fix(auth): redirects no longer send bad auth

zkat committed:
fix(git): integrity hash was failing sometimes

zkat committed:
fix(tests): removing obsolete shrinwkrap tests

zkat committed:
fix(deps): make sure to catch errors here

zkat committed:
fix(shrinkwrap): return shrinkwrap data even if renaming

zkat committed:
fix(tests): update override-bundled

zkat committed:
fix(unpublish): stop clearing caches for now

zkat committed:
move pending tests out of the way

zkat committed:
remove doctor cache dir

zkat committed:
fix indent too

zkat committed:
move away more borked tests

zkat committed:
no need to call shrinkwrap

zkat committed:
bump deps

zkat committed:
more renaming

simonua commented:
@zkat I'm considering recommending `npm cache clean` to our teams prior to upgrading to npm5 to clean out the old pre-5 caches. Is there a better, more suggested way? An npm upgrade to 5 would not automatically wipe out those artifacts, would it?

zkat commented:
@simonua I would even recommend you keep the npm4 cache around until you've confirmed npm5 is currently stable enough on your own project. But cache clear is the way to go once you're ready -- you can also manually preserve `~/.npm/_cacache` and delete everything else to only keep the npm5 cache

zkat commented:
@alanhogan they have been deprecated, and their semantics have been limited due to the design of the underlying cache (which works entirely through http cache rules). The reason they haven't been removed is so that anyone who had things like `--cache-min=999999` still work (it's essentially an alias for `--prefer-offline`. Likewise, `--cache-max=0` is `--prefer-online`.

Furthermore, we consider the _act_ of deprecation to be a breaking change, as a matter of policy. Deprecations might just not happen to break CI (they might, sometimes, if there's new `warning` output, though).

zkat commented:
This landed in `npm@5.0.0` 🎉

alanhogan commented:
Excellent explanation, thank you.

rally25rs commented:
Was `cache ls` intentionally removed during this cache rewrite? Is there a replacement command?

```
$ npm --version
5.0.0

$ npm cache ls
npm ERR! Usage: npm cache add <tarball file>
npm ERR! npm cache add <folder>
npm ERR! npm cache add <tarball url>
npm ERR! npm cache add <git url>
npm ERR! npm cache add <name>@<version>
npm ERR! npm cache clean
npm ERR! npm cache verify
```

If so, docs should probably be updated now that v5 is out
https://docs.npmjs.com/cli/cache

zkat commented:
@rally25rs https://github.com/npm/npm/blob/latest/doc/cli/npm-cache.md the docs are updated, but there's issues with the auto-deploy scripts that push them to the site. The removal was intentional: the cache structure changed so much that it's gonna take a bit before there can be a replacement.
